neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+            "num_noise_variables": self.num_noise_variables,
+            "masses": [mass.describe() for mass in self],
+        }
+
+    def __len__(self):
+        """
+        Get length.
+        """
+        return len(self.masses)
+
+    def __getitem__(self, index):
+        """
+        Get item.
+        """
+        return self.masses[index]
+
+    @property
+    def state_variable_names(self):
+        state_var_names = sum([mass.state_variable_names for mass in self], [])
+        # if state variables are not unique per node, append mass label to them
+        if len(state_var_names) != len(set(state_var_names)):
+            state_var_names = [
+                f"{state_var}_{mass.mass_type}" for mass in self for state_var in mass.state_variable_names
+            ]
+        return [state_var_names]
+
+    @property
+    def max_delay(self):
+        return 0.0
+
+    def get_nested_parameters(self):
+        """
+        Return nested dictionary with parameters from all masses within this
+        node.
+
+        :return: nested dictionary with all parameters
+        :rtype: dict
+        """
+        assert self.initialised
+        node_key = f"{NODE_NAME_STR}_{self.index}"
+        nested_dict = {node_key: {}}
+        for mass in self:
+            mass_key = f"{MASS_NAME_STR}_{mass.index}"
+            nested_dict[node_key][mass_key] = mass.parameters
+        return nested_dict
+
+    def init_node(self, **kwargs):
+        """
+        Initialise node and all the masses within.
+
+        :kwargs: optional keyword arguments to init_mass
+        """
+        # we need to have the index already assigned
+        assert self.index is not None
+        # initialise possible sync variables
+        self.sync_symbols = {
+            f"{symbol}_{self.index}": se.Symbol(f"{symbol}_{self.index}") for symbol in self.sync_variables
+        }
+        for mass in self:
+            mass.init_mass(**kwargs)
+        assert all(mass.initialised for mass in self)
+        self.initialised = True
+
+    def update_parameters(self, parameters_dict):
+        """
+        Update parameters of the node, i.e. recursively update all the masses.
+
+        :param parameters_dict: new parameters for this node, same format as
+            `get_nested_parameters`, i.e. nested dict
+        :type parameters_dict: dict
+        """
+        for mass_key, mass_params in parameters_dict.items():
+            if MASS_NAME_STR in mass_key:
+                mass_index = int(mass_key.split("_")[-1])
+                assert mass_index == self.masses[mass_index].index
+                self.masses[mass_index].update_parameters(mass_params)
+            else:
+                logging.warning(f"Not sure what to do with {mass_key}...")
+
+    @staticmethod
+    def _strip_index(symbol_name):
+        """
+        Strip index value from the symbol name.
+        """
+        return "_".join(symbol_name.split("_")[:-1])
+
+    def _callbacks(self):
```

Comment:
ignore `callbacks` and `numba_callbacks` for now - they are important for AdEx but not otherwise